mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+ *
+ * </pre>
+ *
+ * If the document is modified by another thread, that update will not
+ * appear as an update in the result. But, we do not know which
+ * documents were failed to update this way, only the number of
+ * documents actually updated. So, we read the database again to see
+ * which docvers have been changed. Another thread may come an modify
+ * the documents before this step but after the modifications are
+ * applied, meaning we may find out more documents have been modified
+ * than initially known. Here the question is which documents have our
+ * modifications, and which don't. We would like to filter out the
+ * documents that don't have our modifications, and mark them as
+ * concurrent update failures, because any update performed after our
+ * update is valid.
+ *
+ * This is important for incremental updates. Example:
+ * <pre>
+ *   concurrent update failure:
+ *      thread1:                        thread2:
+ *      read doc, ver=1, field=1
+ *                                read doc, ver=1, field=1
+ *      set field=2
+ *      submit update with ver=2
+ *                                set field=2
+ *                                submit update with ver=2
+ *                                perform db update
+ *      update fails, ver!=1
+ * </pre>
+ *
+ * In the above scenario, the increment operation for thread1 should
+ * fail, because thread2 modified the document after thread1 read
+ * it. This is returned as one missing document in the write result of
+ * thread 1. At this point, thread1 doesn't know which document
+ * failed, so it reads the db to find out:
+ * <pre>
+ *   thread1:                            thread2:
+ *                                     modify another doc, ver=2, set ver=3
+ *   read all docs with ver=2
+ *   
+ * </pre>
+ *
+ * At this point, thread1 sees two documents missing, instead of
+ * one. The first document doesn't have thread1 updates, the second
+ * document has thread1 updates, but thread2 applied its own updates
+ * before thread1 can find out which document was modified. Returning
+ * the second document as failed might result in the client retrying
+ * the update, applying the changes twice,
+ *
+ * To prevent this scenario, we will keep the document versions in an
+ * array, so we'll know which updates are applied to the document. To
+ * prevent unbounded array growth, we will truncate the array using
+ * the docver timestamp.
+ *
+ */
+
+
+
+public class MongoSafeUpdateProtocol {
+
+    private static final Logger LOGGER=LoggerFactory.getLogger(MongoSafeUpdateProtocol.class);
+    
+    public static final String DOCVER="docver";
+
+    public static final String DOCVER_FLD=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER;
+    public static final String DOCVER_FLD0=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER+".0";
+
+    private static final long TOO_OLD=10l*60l*1000l; // Any docver older than 10 minutes is to old
```

Comment:
Renamed constant, and changed it to 1 min.

Your last point about not finding doc at the boundaries of TOO_OLD: someone
must have modified the doc for it to be not found, so it is still a
legitimate concurrent update error.

On Fri, Aug 26, 2016 at 1:03 PM, Alec Henninger <notifications@github.com>
wrote:

> In mongo/src/main/java/com/redhat/lightblue/mongo/crud/
> MongoSafeUpdateProtocol.java
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303#discussion_r76471105>
> :
>
> > + * the docver timestamp.
> > + *
> > + */
> > +
> > +
> > +
> > +public class MongoSafeUpdateProtocol {
> > +
> > +    private static final Logger LOGGER=LoggerFactory.getLogger(MongoSafeUpdateProtocol.class);
> > +
> > +    public static final String DOCVER="docver";
> > +
> > +    public static final String DOCVER_FLD=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER;
> > +    public static final String DOCVER_FLD0=Translator.HIDDEN_SUB_PATH.toString()+"."+DOCVER+".0";
> > +
> > +    private static final long TOO_OLD=10l*60l*1000l; // Any docver older than 10 minutes is to old
>
> IMHO consider naming TOO_OLD_MS (I'm a stickler for putting units in var
> names when it's a duration) or use JDK8 Duration type.
>
> It sounds like our constraints on the time are this: we want to balance
> not having docVer array too large, but giving enough time between our
> update and our concurrent mod check so that it finds our current
> transactions docVer in the array before it is blown away. Correct?
>
> If so, then probably a minute or two is okay amount of time for us to find
> concurrent mods. It would probably depend on client's mongo and timeout
> configuration, though. For example, if a long update starts with MMAPv1
> just after the commit, then depending on timeout configuration, maybe the
> collection would be locked for a while before the find concurrent updates
> could run, and we would miss it if another transaction came through and
> cleaned up the old version before our check finished.
>
> Of course, this would be rare, cause lots of other problems (if collection
> locked for so long), and the side effect would be clients seeing a false
> positive of concurrent update (not world ending, but not good).
>
> Maybe there is another failure condition -- when
> findConcurrentModifications too close to TOO_OLD? Then we cannot be certain
> that docs not found by check were not found because of concurrent mod, or
> because their docVer was removed before we could check.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/lightblue-platform/lightblue-mongo/pull/303/files/6e592caaaf4c79fd821914fdb9640896fa145ab0#r76471105>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADgDDchuH8Jfymlk0JJYZncC4nb4yATPks5qjzidgaJpZM4JtbtZ>
> .
>